applications/StructuralMechanicsApplication/custom_advanced_constitutive/plane_stress_d_plus_d_minus_damage_masonry_2d.cpp

```diff
@@ -7,752 +7,1076 @@
+			ThresholdTension = UniaxialStressTension;
+		this->CalculateDamageTension(data, ThresholdTension, DamageParameterTension);
+
+		if(UniaxialStressCompression > ThresholdCompression)
+			ThresholdCompression = UniaxialStressCompression;
+		this->CalculateDamageCompression(data, ThresholdCompression, DamageParameterCompression);
+
+		TemporaryImplicitThresholdTension = ThresholdTension;
+		TemporaryImplicitThresholdTCompression = ThresholdCompression;
+	}
+
+	// calculation of stress tensor
+	noalias(PredictiveStressVector)  = (1.0 - DamageParameterTension)     * data.EffectiveTensionStressVector;
+	noalias(PredictiveStressVector) += (1.0 - DamageParameterCompression) * data.EffectiveCompressionStressVector;
+}
+/***********************************************************************************/
+/***********************************************************************************/
+void DamageDPlusDMinusMasonry2DLaw::CheckDamageLoadingUnloading(
+	bool& is_damaging_tension, 
+	bool& is_damaging_compression)
+{
+	const double F_tension 		= UniaxialStressTension - CurrentThresholdTension;
+	const double F_compression 	= UniaxialStressCompression - CurrentThresholdCompression;
+
+	is_damaging_tension = false;
+	if (F_tension > 0.0){
+		is_damaging_tension = true;
+	}
+	is_damaging_compression = false;
+	if (F_compression > 0.0){
+		is_damaging_compression = true;
+	}
+}
+/***********************************************************************************/
+/***********************************************************************************/
+void DamageDPlusDMinusMasonry2DLaw::CalculateTangentTensor(
+	Parameters& rValues, 
+	Vector StrainVector,
+	Vector PredictiveStressVector,
+	CalculationData& data,
+	Properties props)
+{
+	// prepare constitutive matrix
+	Matrix& constitutive_matrix = rValues.GetConstitutiveMatrix();
+	if(constitutive_matrix.size1() != VoigtSize || constitutive_matrix.size2() != VoigtSize)
+		constitutive_matrix.resize(VoigtSize, VoigtSize);
+
+	// save internal variables
+	double save_threshold_tension			= ThresholdTension;
+	double save_threshold_compression		= ThresholdCompression;
+	double save_damage_tension				= DamageParameterTension;
+	double save_damage_compression			= DamageParameterCompression;
+	double save_uniaxial_stress_tension		= UniaxialStressTension;
+	double save_uniaxial_stress_compression	= UniaxialStressCompression;
+
+	// perturbation parameter
+	double perturbation_factor = 1.0E-8;
+
+	// perturbed vectors
+	Vector perturbated_strain_vector(VoigtSize);
+	Vector perturbated_stress_vector(VoigtSize);
+
+	// apply perturbation to each strain component...
+	for(size_t j = 0; j < VoigtSize; j++)
+	{
+		noalias(perturbated_strain_vector) = StrainVector;
+
+		perturbated_strain_vector(j) = StrainVector(j) + perturbation_factor;
+		this->CalculateMaterialResponseInternal(perturbated_strain_vector, perturbated_stress_vector, data, props);
+
+		for(size_t i = 0; i < VoigtSize; i++)

```

Comment:
you could use our tangent_operator_utility?

Reply:
I have to store some internal variables inside the computation of the tangent matrix (due to IMPLEX integration), that is why I wrote it in there.